Added D1 Regression Metrics

anmolsjoshi opened the pull request:
Related to #307 

Description:
Adds D1 regression metrics.

Check list:
* [x] New tests are added. 
* [x] New doc strings: text and/or example code are in RST format
* [x] Documentation is updated (if required)

vfdev-5 commented:
@anmolsjoshi I edited your message : `Fixes` -> `Related to` as this PR fixes partly the issue and if we keep Fixes when PR is accepted, the issue will be closed.

vfdev-5 commented:
@anmolsjoshi could you put the code into contrib please.

anmolsjoshi commented:
@vfdev-5 metrics have been moved to contrib and tests have been added. 

For Fractional Bias, I was thinking of adding a test to check if any element of y_pred + y is 0. Thoughts?

docstrings need to be updated as in #331 - will commit that later tonight.

vfdev-5 commented on ignite/contrib/metrics/fractional_bias.py:
```diff
@@ -0,0 +1,47 @@
+
+    More details can be found in `here`_.
+
+    :math:`\text{FB} = \frac{1}{n}\sum _j^n\frac{2 * (A_j - P_j}{A_j + P_j})`
```
```suggestion
    :math:`\text{FB} = \frac{1}{n}\sum _j^n\frac{2 * (A_j - P_j)}{A_j + P_j}`
```

anmolsjoshi replied:
Done!

vfdev-5 commented on ignite/contrib/metrics/mae.py:
```diff
@@ -4,11 +4,25 @@
+
+    More details can be found in `here`_.
+
+    :math:`\text{MaxAE} = \max_{j=1,n} \left(|A_j-P_j\right|)`
```
```suggestion
    :math:`\text{MaxAE} = \max_{j=1,n} \left( \lvert A_j-P_j \rvert \right)`
```

anmolsjoshi replied:
Done!

vfdev-5 commented on ignite/contrib/metrics/mae.py:
```diff
@@ -4,11 +4,25 @@
 
 
 class MaximumAbsoluteError(Metric):
```
Could you please also rename the file `mae.py` -> `maximum_absolute_error.py` to have a coherence with others.

anmolsjoshi replied:
Done!

anmolsjoshi commented:
@vfdev-5 all comments have been incorporated. 

- docstrings have been updated as requested.
- mean_percentage_error metric and tests have been deleted. 
- regression module has been added, containing all regression related metrics. 
- mae has been renamed as maximum_absolute_error.
- all the tests have been updated appropriately.

vfdev-5 commented:
@jasonkriss or @alykhantejani review and merge please

zasdfgbnm commented on tests/ignite/contrib/metrics/regression/test_fractional_bias.py:
```diff
@@ -0,0 +1,46 @@
+    ground_truth = np.random.randn(4)
+
+    m = FractionalBias()
+    m.reset()
```
No need to do `reset` here, it would be called automatically by `__init__`. As in https://github.com/pytorch/ignite/blob/master/ignite/metrics/metric.py#L23

Same for other metrics.

anmolsjoshi replied:
Incorporated this comment.

zasdfgbnm commented on tests/ignite/contrib/metrics/regression/test_manhattan_distance.py:
```diff
@@ -0,0 +1,38 @@
+import pytest
+
+
+def test_zero_div():
```
There is no div at all in `ManhattanDistance`, so this shouldn't be called `test_zero_div`

anmolsjoshi replied:
Please see comment above about removing NotComputableError entirely.

zasdfgbnm commented on ignite/contrib/metrics/regression/fractional_bias.py:
```diff
@@ -0,0 +1,45 @@
+        denominator = y_pred + y.view_as(y_pred)
+
+        if (denominator == 0).any():
+            raise NotComputableError('The denominator has 0.')
```
@vfdev-5 for this case, different from https://github.com/pytorch/ignite/pull/331, should we just leave a warning and return an `inf` or `nan` whichever the result is?  The comment in https://github.com/pytorch/ignite/pull/331#issuecomment-437721992 no longer applies, because here the `y_pred` is the value of model output, which is not under the control of user's hand. During training, `y_pred` could be any ridiculous number. Users might be interested in this metric (because when the model is good enough, this is well-defined), and doesn't care if occasionally this is ill-defined during training. In this case, raising `NotComputableError` might be annoying to the user because this stops their training.

vfdev-5 replied:
@zasdfgbnm I agree, thanks for pointing on this ! Let's use warnings.warn('The denominator is 0. Error becomes NaN') and set `_sum_of_errors` to np.nan.

zasdfgbnm replied:
@vfdev-5 how about `inf`? Would it be better to leave it an `inf` if denominator is 0 but numerator is not?

vfdev-5 replied:
Mathematically correct would be plus or minus infinity. Maybe we can do similarly to torch:
```
import torch

torch.ones(2) / torch.eye(2)

tensor([[1., inf],
        [inf, 1.]])
```

vfdev-5 replied:
> I feel here we are a bit over-worried about divide by zero case. 

Yes, probably this is the best path to take :) Let's see others opinion. But I think you are right, we probably should just remove this `raise`.

anmolsjoshi replied:
I agree with setting this as inf and removing the raise error.

vfdev-5 commented:
@zasdfgbnm thanks for the review !

> Also, @vfdev-5 what is the best way to reference the paper in docstring? My personal opinion is this way: Botchkarev 2018, that is use name + year as link text, and use the abstract page of arXiv as link. Whichever we choose, we should be consistent in #331 as well. cc: @weihuangxu

I have no special preference for this, `Botchkarev 2018` seems ok for me. The problem could be with with sphinx docs rendering. Maybe we can do the same as in PyTorch, for example, [citation of ADADELTA](https://pytorch.org/docs/stable/_modules/torch/optim/adadelta.html#Adadelta).

vfdev-5 commented on ignite/contrib/metrics/regression/manhattan_distance.py:
```diff
@@ -0,0 +1,39 @@
+    """
+    def reset(self):
+        self._sum_of_errors = 0.0
+        self._num_examples = 0
```
@zasdfgbnm correct, unused variable, thanks for catching this !

zasdfgbnm replied:
@vfdev-5 What is your opinion on the second part of my comment? i.e. about

> All metrics that has not been updated yet after last reset should raise `NotComputableError` when `compute`

zasdfgbnm replied:
@vfdev-5 here for this case, instead of using `_num_examples`, it might be better to use a boolean variable called `_updated`, or even just remove the check (I don't know which choice is better).

vfdev-5 replied:
@zasdfgbnm this is a correct assumption. How would you implement this ?
A resetable counter of `update` method that is checked in `compute`... :|

zasdfgbnm replied:
@vfdev-5 My suggestion is:
1. For the purpose of this PR, just remove `_num_examples` and don't worry about not being updated. If we are worried about this metric, then we should also be worried about more other metrics, so it worth a separate PR to correct all of them.
2. If we are using metrics only with engines, then it should be safe to assume a metric will never be computed before updated.
3. If we still want to check if it is updated before compute, some change in `Metric` like this might work:

```python
class Metric:
    def  __init__(self, ...):
        def wrapped_reset():
            self._updated = False
            self.reset()
        self.reset = functool.wraps(self.reset)(wrapped_reset)

        def wrapped_update(output):
            self._updated = True
            self.update(output)
        self.update =  functool.wraps(self.update)(wrapped_update)

        def wrapped_compute():
            if not self._updated:
                raise NotComputableError('not updated before compute')
            return self.compute()
        self.compute =  functool.wraps(self.compute)(wrapped_compute)
```

anmolsjoshi replied:
For ManhattanDistance, since there is no division, it might be best to just return whatever the current value of self._sum_of_errors is. I have removed NotComputableError for this metric.

vfdev-5 commented on tests/ignite/contrib/metrics/regression/test_manhattan_distance.py:
```diff
@@ -0,0 +1,38 @@
+import pytest
+
+
+def test_zero_div():
```
```suggestion
def test_no_elements():
```

anmolsjoshi replied:
Please see comment above.

anmolsjoshi commented:
@vfdev-5 @zasdfgbnm thanks for the review and discussion! Please see me latest commit.

- removed m.reset() from all the tests per this [comment](https://github.com/pytorch/ignite/pull/327#discussion_r233275118)
- changed docstring to simple author + year, with link to abstract per this [comment](https://github.com/pytorch/ignite/pull/327#issuecomment-438498035)
- removed raise error of denominator == 0 in fractional_bias per this [comment](https://github.com/pytorch/ignite/pull/327#discussion_r233286181)
- removed manhattan_distance test and raise error for 0 iterations per this [comment](https://github.com/pytorch/ignite/pull/327#discussion_r233265126)

vfdev-5 commented:
@jasonkriss or @alykhantejani please review and merge if OK

anmolsjoshi commented:
@vfdev-5 @zasdfgbnm please note that I have updated the metrics to calculate e = A - P. 

Last few commits had the opposite, tests have been updated also.

jasonkriss commented:
@anmolsjoshi Can you resolve the conflicts? Sorry about that. We'll get this merged as soon as that happens.

anmolsjoshi commented:
@jasonkriss I have resolved the conflict. let me know if there are any issues.

jasonkriss commented:
Thanks @anmolsjoshi!
